server/mscalendar/availability.go

```diff
@@ -155,14 +134,135 @@ func (m *mscalendar) retrieveUsersToSync(userIndex store.UserIndex, syncJobSumma
+
+	numberOfLogs := 0
+	users := []*store.User{}
+	calendarViews := []*remote.ViewCalendarResponse{}
+	for _, u := range userIndex {
+		user, err := m.Store.LoadUser(u.MattermostUserID)
+		if err != nil {
+			syncJobSummary.NumberOfUsersFailedStatusChanged++
+			if numberOfLogs < logTruncateLimit {
+				m.Logger.Warnf("Not able to load user %s from user index. err=%v", u.MattermostUserID, err)
+			} else if numberOfLogs == logTruncateLimit {
+				m.Logger.Warnf(logTruncateMsg)
+			}
+			numberOfLogs++
+			continue
+		}
+
+		// If user does not have the proper features enabled, just go to the next one
+		if !(user.Settings.UpdateStatus || user.Settings.ReceiveReminders) {
+			continue
+		}
+
+		users = append(users, user)
+	}
+
+	if len(users) == 0 {
+		return users, calendarViews, errNoUsersNeedToBeSynced
+	}
+
+	in := make(chan store.User)
+	out := make(chan StatusSyncJobSummary)
+	finishChan := make(chan struct{})
+	var wg sync.WaitGroup
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	// Start workers
+	for i := 0; i <= concurrency; i++ {
+		wg.Add(1)
+		go func(m mscalendar, c context.Context, w *sync.WaitGroup, in chan store.User, out chan StatusSyncJobSummary) {
+			defer w.Done()
+			for {
+				select {
+				case <-c.Done():
+					m.Logger.Errorf("Timeout processing users availability")
+					break
+				case user, ok := <-in:
+					if !ok {
+						// Closed channel
+						return
+					}
+
+					js := StatusSyncJobSummary{}
+					engine, err := m.FilterCopy(withActingUser(user.MattermostUserID))
+					if err != nil {
+						m.Logger.Warnf("Not able to enable active user %s from user index. err=%v", user.MattermostUserID, err)
+						continue
+					}
+
+					calendarUser := newUserFromStoredUser(&user)
+					js.CalendarEvents, err = engine.GetCalendarEvents(calendarUser, start, end, true)
+					if err != nil {
+						js.NumberOfUsersFailedStatusChanged++
+						m.Logger.With(bot.LogContext{
+							"user": user.MattermostUserID,
+							"err":  err,
+						}).Errorf("error getting calendar events")
+						continue
+					}
+
+					out <- js
+				}
+			}
+		}(*m, ctxTimeout, &wg, in, out)
+	}
+
+	// Populate the input channel with the users and wait for the workers to finish
+	go func(users []*store.User, in chan store.User, out chan StatusSyncJobSummary) {
+		for _, user := range users {
+			in <- *user
+		}
+		close(in)
+
+		wg.Wait()
+		close(out)
+		close(finishChan)
+	}(users, in, out)
+
+	// Read results and wait until all workers have finished.
+	for {
+		select {
+		case js, ok := <-out:
+			if !ok {
+				continue
+			}
+
+			syncJobSummary.NumberOfUsersFailedStatusChanged += js.NumberOfUsersFailedStatusChanged
+			calendarViews = append(calendarViews, js.CalendarEvents)
+		case <-finishChan:
+			if len(calendarViews) == 0 {
+				return users, calendarViews, fmt.Errorf("no calendar views found")
+			}
+
+			return users, calendarViews, nil
+		}
```

Comment:
Why is this case needed? Doesn't closing `out` already indicate that there isn't more to read?

Reply:
it was pure cosmetic preference, didn't want to put all the logic in the same `case` statement, but it makes sense.